core/trino-main/src/main/java/io/trino/operator/output/PositionsAppenderPageBuilder.java

```diff
@@ -144,6 +145,31 @@ public boolean isEmpty()
         return declaredPositions == 0;
     }
 
+    public Optional<Page> maybeFlushBeforeRelease()
+    {
+        if (declaredPositions == 0) {
+            return Optional.empty();
+        }
+        boolean forceFlush = false;
+        for (UnnestingPositionsAppender positionsAppender : channelAppenders) {
+            if (positionsAppender.shouldForceFlushBeforeRelease()) {
+                // dictionary encoding will be preserved, so force the current page to be flushed
+                forceFlush = true;
+                break;
+            }
+            else {
+                // transition from dictionary to direct mode if necessary, since we won't be able to reuse the
+                // same dictionary from the new operator
+                positionsAppender.flattenPendingDictionary();
```

Comment:
Those dictionaries are going to end up flattened anyways during compaction as part of serialization, but this does relate to the correctness of the output bytes reported (technically, this is more correct but still wrong because all dictionary outputs are reported incorrectly).